Improve iso-manager script error handling

alexandre-allard opened the pull request:
**Component**: scripts

**Context**: 

**Summary**:

**Acceptance criteria**: 


---

Closes: #3079
Closes: #2862

alexandre-allard committed:
scripts: get archive absolute path

We need the absolute path to put it in the
bootstrap.yaml configuration file, otherwise
Salt will not be able to retrieve the archive
when called.

alexandre-allard committed:
scripts: use common library functions

Instead of redeclaring some functions, use
the common.sh library

bert-e commented:
# Hello alexandre-allard-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast

TeddyAndrieux commented on scripts/iso-manager.sh:
```diff
@@ -139,14 +80,29 @@ _set_env() {
+        sed -rn 's/^- (.+)$/\1/p'
+}
+
+_check_archives() {
+    local -a not_existing
+
+    for archive; do
+        if [ ! -e "$archive" ]; then
+            not_existing+=("$archive")
+        fi
+    done
+
+    if (( ${#not_existing[@]} )); then
+        echo "The following archives do not exist:" >&2
+        for archive in "${not_existing[@]}"; do
+            echo "- $archive" >&2
+        done
+        return 1
+    fi
+
+    return 0
 }
```
IMHO this check should be in salt so that we can also check this in bootstrap, also this remind me this ticket #2862

alexandre-allard replied:
I agree, but having it here allow to fail early before touching this file.
Maybe we need it here and in Salt too or we could also do like we did for Solutions, not using file.serialize and having a custom module adding/removing ISO from this file.

TeddyAndrieux replied:
The salt module could take archive list as input (and default to pillar content)

Hmm yes maybe it make sense to have a dedicated Salt module to add iso to bootstrap file

TeddyAndrieux commented on scripts/iso-manager.sh:
```diff
@@ -157,21 +113,15 @@ _add_archives() {
-    IFS=" " read -r -a \
-        archives <<< "$(salt-call pillar.get metalk8s:archives \
-        --out txt | cut -d' ' -f2- | tr -d '[],')"
+    local -a archives=(${CONFIGURED_ARCHIVES[@]+"${CONFIGURED_ARCHIVES[@]}"})
```
I'm not sure I understand what it does exactly
```
${CONFIGURED_ARCHIVES[@]+"${CONFIGURED_ARCHIVES[@]}"}
```

alexandre-allard replied:
Same as:
```
local -a archives
if (( ${#CONFIGURED_ARCHIVES[@]} )); then
    archives=("${CONFIGURED_ARCHIVES[@]}")
fi
```
I can change it if you feel it's not clear enough.

TeddyAndrieux replied:
Ok, as you want now that I know this, I'm ok to keep this one line :smiley:

alexandre-allard replied:
Anyway, this line no longer exist :'( it's in Salt module now.

alexandre-allard committed:
salt: Add function to add/remove archive from bootstrap config

Refs: #3079

alexandre-allard committed:
scripts: Use salt function to add archive to bootstrap config

This way we get rid of shell logic and we can
check for archive validity.

Refs: #3079

TeddyAndrieux commented on salt/_modules/metalk8s.py:
```diff
@@ -214,26 +239,8 @@ def get_archives(archives=None):
+        info = archive_info_from_product_txt(archive)
+        env_name = 'metalk8s-{0}'.format(info['version'])
 
         # Warn if we have 2 archives with the same version
         if env_name in res:
```
I think we may want to raise in this case as well (maybe in another PR/ticket)

alexandre-allard replied:
Yes, I was wondering if we wanted to do that, I can handle it here.

TeddyAndrieux commented on salt/_modules/metalk8s.py:
```diff
@@ -615,3 +622,36 @@ def get_from_map(value, saltenv=None):
+        with open(BOOTSTRAP_CONFIG, 'r') as fd:
+            config = yaml.safe_load(fd)
+    except IOError as exc:
+        msg = 'Failed to load bootstrap config file at "{}": {}'.format(
+            BOOTSTRAP_CONFIG, str(exc)
+        )
+        raise CommandExecutionError(message=msg)
```
Not link to this PR but in Python3 it's better to raise exception and use `from exc` to have both exceptions (check this https://www.python.org/dev/peps/pep-3134/#explicit-exception-chaining)

NOTE: I fix all other except in my pylint PR #3066 (e.g.: https://github.com/scality/metalk8s/pull/3066/commits/e4557955353e02b360fe71926dbf1760d5ce5cc1)

alexandre-allard replied:
Oh, didn't know that, I will change it.
Thanks for the tip.

TeddyAndrieux commented on salt/_modules/metalk8s.py:
```diff
@@ -615,3 +622,36 @@ def get_from_map(value, saltenv=None):
     )
+
+
+def configure_archive(archive, removed=False):
```
Minor: to me it's more "remove" than "removed" no ?

TeddyAndrieux commented on salt/_modules/metalk8s.py:
```diff
@@ -615,3 +622,36 @@ def get_from_map(value, saltenv=None):
+        )
+        raise CommandExecutionError(message=msg)
+
+    if removed:
+        try:
+            config['archives'].remove(archive)
+        except ValueError:
+            pass
+    else:
+        if archive not in config['archives']:
+            config['archives'].append(archive)
```
Minor: To me it could be cool to have a simple message output like `<archive> added` `<archive> removed` `<archive> already present` `<archive> already absent` something like that

TeddyAndrieux reviewed (commented):
Few comments otherwise LGTM (just unit tests are missing + likely a changelog entry)

alexandre-allard committed:
salt: Raise if we have duplicated MetalK8s archive versions

alexandre-allard committed:
salt: fix assertDictContainsSubset deprecation warning

alexandre-allard committed:
salt: Add unit tests for `metalk8s.configure_archive`

Refs: #3079

alexandre-allard committed:
Add changelog entry for iso-manager hardening

alexandre-allard committed:
scripts: Check bootstrap config archives in iso-manager

Check that all configured archives are valid path
pointing to valid MetalK8s archives, to fail early.

Refs: #3079

TeddyAndrieux reviewed (dismissed):
LGTM

alexandre-allard commented:
/approve

bert-e commented:
# Branches have diverged



This pull request's source branch `bugfix/3079-iso-manager-improve-error-message` has diverged from
`development/2.8` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.8` into `bugfix/3079-iso-manager-improve-error-message`
* Rebase `bugfix/3079-iso-manager-improve-error-message` onto `origin/development/2.8`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.




*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.8`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.8`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`



Please check the status of the associated issue None.

Goodbye alexandre-allard-scality.
